File: go/web.go
package main

import (
	"bytes"
	"encoding/hex"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"regexp"
	"strings"

	"github.com/sirupsen/logrus"
)

const baseWebUrl = "https://ko0.com"

const submitUrl = "https://ko0.com/submit/"

const userAgent = "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.71 Safari/537.36Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.64 Safari/537.11Mozilla/5.0 (Windows; U; Windows NT 6.1; en-US) AppleWebKit/534.16 (KHTML, like Gecko) Chrome/10.0.648.133 Safari/534.16"

var myUrl string

func InitMyUrl(secretKeyHex string) error {
	rKey := []byte(secretKeyHex)
	rkey, err := GetSha256(rKey)
	if err != nil {
		return err
	}
	rkey, err = GetSha256(rkey)
	if err != nil {
		return err
	}
	rkeyHex := hex.EncodeToString(rkey)
	myUrl = baseWebUrl + "/" + rkeyHex[0:16]
	logrus.Infof("myUrl: %s", myUrl)
	return nil
}

func GetContentFromWeb() ([]byte, error) {
	client, err := Get_client()
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequest("GET", myUrl, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", userAgent)
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	bodyText, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	re := regexp.MustCompile(`class[\s]*=[\s]*"txt_view">[\s]*<p>(.+)<\/p>`)
	match := re.FindStringSubmatch(string(bodyText))
	if len(match) == 0 {
		return nil, fmt.Errorf("can not find content")
	}
	// 读取
	encryptedata, err := hex.DecodeString(match[1])
	if err != nil {
		return nil, err
	}
	// 解密
	decryptedata, err := crypter.Decrypt(encryptedata)
	if err != nil {
		return nil, err
	}
	return decryptedata, nil
}

func PostContentToWeb(context []byte) error {
	encryptedata, err := crypter.Encrypt(context)
	if err != nil {
		return err
	}
	encryptedataHex := hex.EncodeToString(encryptedata)

	client, err := Get_client()
	if err != nil {
		return err
	}
	csrfmiddlewaretoken, err := getPostCsrfmiddlewaretoken(&client)
	if err != nil {
		return err
	}
	payload := &bytes.Buffer{}
	writer := multipart.NewWriter(payload)
	_ = writer.WriteField("csrfmiddlewaretoken", csrfmiddlewaretoken)
	_ = writer.WriteField("txt", encryptedataHex)
	_ = writer.WriteField("code", myUrl[strings.LastIndex(myUrl, "/")+1:])
	_ = writer.WriteField("sub_type", "T")
	_ = writer.WriteField("file", "")
	err = writer.Close()
	if err != nil {
		return err
	}
	req, err := http.NewRequest("POST", submitUrl, payload)
	if err != nil {
		return err
	}
	//fmt.Println(client.Jar.Cookies(req.URL))
	//fmt.Println(writer.FormDataContentType())
	req.Header.Set("Content-Type", writer.FormDataContentType())
	req.Header.Set("Referer", myUrl)
	req.Header.Set("User-Agent", userAgent)
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	bodyText, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if !strings.Contains(string(bodyText), "success") {
		return fmt.Errorf("post content failed")
	}
	return nil
}

func getPostCsrfmiddlewaretoken(client *http.Client) (string, error) {
	req, err := http.NewRequest("GET", myUrl, nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("User-Agent", userAgent)
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	bodyText, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	return parseCsrfmiddlewaretoken(bodyText)
}

func parseCsrfmiddlewaretoken(content []byte) (string, error) {
	re := regexp.MustCompile(`name="csrfmiddlewaretoken" value="(.+)">`)
	match := re.FindStringSubmatch(string(content))
	if len(match) == 0 {
		return "", fmt.Errorf("can not find csrfmiddlewaretoken")
	}
	return match[1], nil
}

File: windSend-rs/src/web.rs
use tracing::debug;

pub static BASE_WEB_URL: &str = "https://ko0.com";
pub static SUBMIT_URL: &str = "https://ko0.com/submit/";
pub static USER_AGENT: &str = "Mozilla/5.0 (Windows NT 6.1; WOW64) \
    AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.71 Safari/537.36Mozilla/5.0 \
    (X11; Linux x86_64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.64 \
    Safari/537.11Mozilla/5.0 (Windows; U; Windows NT 6.1; en-US) AppleWebKit/534.16 \
    (KHTML, like Gecko) Chrome/10.0.648.133 Safari/534.16";

lazy_static::lazy_static!(
    pub static ref MY_URL: String = init_my_url();
);

fn init_my_url() -> String {
    let secret_key_hex = crate::config::GLOBAL_CONFIG
        .read()
        .unwrap()
        .secret_key_hex
        .clone();
    let r_key = secret_key_hex.as_bytes();
    let r_key = crate::utils::encrypt::compute_sha256(r_key);
    let r_key = crate::utils::encrypt::compute_sha256(&r_key);
    let r_key_hex = hex::encode(r_key);
    let my_url = format!("{}/{}", BASE_WEB_URL, &r_key_hex[0..16]);
    tracing::info!("my_url: {}", my_url);
    my_url
}

pub async fn get_content_from_web() -> Result<Vec<u8>, Box<dyn std::error::Error>> {
    let client = reqwest::Client::new();
    let resp = client
        .get(&*MY_URL)
        .header("User-Agent", USER_AGENT)
        .send()
        .await?;
    let body_text = resp.text().await?;
    let re = regex::Regex::new(r#"class[\s]*=[\s]*"txt_view">[\s]*<p>(.+)<\/p>"#)?;
    let matchs = re.captures(&body_text).ok_or("can not find content")?;
    debug!("{:?}", matchs);
    let mut encrypted_data = hex::decode(matchs.get(1).unwrap().as_str())?;
    let decrypt_data = crate::config::get_cipher()?.decrypt(&mut encrypted_data, "".as_bytes())?;
    Ok(decrypt_data.to_vec())
}

pub async fn post_content_to_web(context: &[u8]) -> Result<(), Box<dyn std::error::Error>> {
    let encrypted_data = crate::config::get_cipher()?.encrypt(context, "".as_bytes())?;
    let encrypted_data_hex = hex::encode(&encrypted_data);
    let client = reqwest::ClientBuilder::new()
        .cookie_store(true)
        .build()
        .unwrap();
    let csrfmiddlewaretoken = get_post_csrfmiddlewaretoken(&client).await?;
    let mut payload = reqwest::multipart::Form::new();
    debug!("csrfmiddlewaretoken: {}", csrfmiddlewaretoken);
    let my_url = (*MY_URL).replace('\\', "/");
    let post_code = my_url
        .split('/')
        .next_back()
        .ok_or_else(|| format!("invalid post url: {my_url}"))?
        .to_string();
    debug!("post_code: {}", post_code);
    payload = payload.text("csrfmiddlewaretoken", csrfmiddlewaretoken);
    payload = payload.text("txt", encrypted_data_hex);
    payload = payload.text("code", post_code);
    payload = payload.text("sub_type", "T");
    payload = payload.text("file", "");
    let resp = client
        .post(SUBMIT_URL)
        .header("Referer", &*MY_URL)
        .header("User-Agent", USER_AGENT)
        .multipart(payload)
        .send()
        .await?;
    let body_text = resp.text().await?;
    debug!("{}", body_text);
    if !body_text.contains("success") {
        return Err("post content failed".into());
    }
    Ok(())
}

async fn get_post_csrfmiddlewaretoken(
    client: &reqwest::Client,
) -> Result<String, Box<dyn std::error::Error>> {
    let resp = client
        .get(&*MY_URL)
        .header("User-Agent", USER_AGENT)
        .send()
        .await?;
    let body_text = resp.text().await?;
    parse_csrfmiddlewaretoken(&body_text)
}

fn parse_csrfmiddlewaretoken(content: &str) -> Result<String, Box<dyn std::error::Error>> {
    let re = regex::Regex::new(r#"name="csrfmiddlewaretoken" value="(.+)">"#)?;
    let matchs = re
        .captures(content)
        .ok_or("can not find csrfmiddlewaretoken")?;
    Ok(matchs.get(1).unwrap().as_str().to_string())
}
